### PR TITLE
[TE] rootcause - request scheduler

### PR DESCRIPTION
**Pull request by apucher**
This PR adds a request scheduler service to handle large data sets gracefully. The scheduler supports multi-level priority scheduling to respond to user-requests sensibly while under full load.

### CODE REVIEW

File: thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+//
+/**
+ * Priority Queue via max-heap
+ */
+const top = 0;
+const parent = i => ((i + 1) >>> 1) - 1;
+const left = i => (i << 1) + 1;
+const right = i => (i + 1) << 1;
+
+class PriorityQueue {
```
Comment (jihaozh) [inline thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js]: I like Javascript. Have to implement heap from scratch!

File: thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      starting.pushObject(waiting.pop()); // modified in-place
+    }
+
+    setProperties(this, { running: [...remaining, ...starting] });
+
+    starting.forEach(p => {
+      fetch(p.url)
+        .then(res => { p.result = res; this._update(); })
+        .catch(err => { p.error = err; this._update(); });
+    });
```
Comment (aaronucsd) [inline thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js]: I would swap this forEach for `for..of` and add async/await if you want guarantee for the order of return responses and also if there is dependencies of something here (ex. p.result) in the codes after it. 

`
  async _update() {
..
 for (const p of starting) {
      await fetch(p.url)
        .then(res => { p.result = res; this._update(); })
        .catch(err => { p.error = err; this._update(); });
    });`

File: thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js (hunk 1 above)
Comment (aaronucsd) [inline thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js]: There a need for document for both Classes and also what this new service will deliver. It's good practice on all services and utils etc. So we know how we can use it elsewhere. A link to `stackoverflow` might always give us enough insight. Thanks.

File: thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js (hunk 2 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js]: This is all about concurrency

File: thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js (hunk 1 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js]: ok added more JS doc

**Pull request merged**